### DISCUSSION TITLE
nlohmann on Ubuntu 18 and Ubuntu 20

### DISCUSSION THREAD

**Discussion by sky0193** [nlohmann/json#3712] [Q&A]
Hallo.

I have Code that uses nlohmann json library and i want to compile it for Ubuntu 20 as well as for Ubuntu 18.

So in CMakeList.txt exists a line like `find_package(nlohmann_json REQUIRED)`

I use for both gitlab CI and install it using:  
```
- apt-get update && DEBIAN_FRONTEND="noninteractive" TZ="Europe/Berlin"  apt-get install -y build-essential cmake libmariadb-dev-compat libmariadb-dev pkg-config nlohmann-json3-dev
```

There is no difference in CMakeList.txt neither in installation. In both cases the nlohmann-json3-dev is installed. 
UBUNTU20 WORKS!!!
And when looking on Ubuntu 20 after installation there exists  
```
/usr/lib/cmake/nlohmann_json/nlohmann_jsonConfig.cmake
/usr/lib/cmake/nlohmann_json/nlohmann_jsonConfigVersion.cmake
/usr/lib/cmake/nlohmann_json/nlohmann_jsonTargets.cmake
```
in Ubuntu 18 this files do not exist.
and i get the error:
```
CMake Error at xya/xyv/CMakeLists.txt:3 (find_package):
  By not providing "Findnlohmann_json.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "nlohmann_json", but CMake did not find one.
  Could not find a package configuration file provided by "nlohmann_json"
  with any of the following names:
    nlohmann_jsonConfig.cmake
    nlohmann_json-config.cmake
  Add the installation prefix of "nlohmann_json" to CMAKE_PREFIX_PATH or set
  "nlohmann_json_DIR" to a directory containing one of the above files.  If
  "nlohmann_json" provides a separate development package or SDK, be sure it
  has been installed.
```

if i manually ad the files to `/usr/lib/cmake/nlohmann_json/` the error dissapear. But it feals wrong.
What am i doing wrong? What is the correct way to fix the issue.

**Comment by falbrechtskirchinger** [comment]
https://packages.ubuntu.com/bionic/nlohmann-json-dev

**Reply by sky0193** [reply]
I habe used features of nlohmann-json3-dev. So in case of nlohmann-json-dev it will not compile
and nlohmann-json3-dev also works well for ubuntu18 beside the fact, that the cmake files seems to missing in my case (in case of pakage install of nlohmann-json3-dev )

**Reply by falbrechtskirchinger** [reply]
My point is, where are you getting `nlohman-json3-dev` from? I don't see an official package for 18.xx.

https://packages.ubuntu.com/search?keywords=nlohmann-json3-dev&searchon=names

See https://json.nlohmann.me/integration/ for alternative ways to integrate the library if the version you need is not packaged for your distribution.

**Reply by sky0193** [reply]
i use on Ubuntu 18 apt-get install  nlohmann-json3-dev and it works, so it is not manually installation.
if i google i also find this https://ubuntu.pkgs.org/18.04/kodi-amd64/nlohmann-json3-dev_3.1.2-2~bionic_all.deb.html

**Reply by falbrechtskirchinger** [reply]
That's a third-party repository.

At the end of the day, what do you want "us" to do here? As you've already identified, your package does not contain the required CMake config files. I've pointed you to a resource listing other ways of integrating the library into your project. What else are you looking for?

**Reply by sky0193** [reply]
thank you for your answer, i was not aware, that  nlohmann-json3 is not supported for ubuntu 18, ant the way that goes is install nlohmann-json or add cmake config files manually. Correct me if I misunderstood you.
The only thing that remains a mystery to me is why it is available to install nlohmann-json3 using apt-get install if this is not supported

**Reply by falbrechtskirchinger** [reply]
You got it.

You must have added an APT repository that provides `nlohmann-json3-dev`. Possibly the one you linked to above (`kodi`). :man_shrugging:

**Reply by sky0193** [reply]
thanks a lot!
